### PR TITLE
Call `compute_locs` once per rule

### PR DESCRIPTION
**Pull request by nnethercote**
This fixes the small regressions on `wg-grammar` and `hyper-0.14.18` seen in #95555.

r? @petrochenkov

### CODE REVIEW

**Comment by nnethercote** [comment]
Local instruction count measurements on `check full` builds. Those marked with `*` are from rustc-perf.

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
num-derive-0.3.3 | check | full | -1.99% | 9.93x
*html5ever-0.26.0 | check | full | -1.93% | 9.63x
scroll_derive-0.11.0 | check | full | -1.88% | 9.40x
pest_generator-2.1.3 | check | full | -1.86% | 9.32x
ctor-0.1.21 | check | full | -1.78% | 8.92x
mockall_derive-0.11.0 | check | full | -1.62% | 8.12x
yansi-0.5.0 | check | full | -1.60% | 8.01x
futures-macro-0.3.19 | check | full | -1.55% | 7.77x
serde_derive-1.0.136 | check | full | -1.32% | 6.59x
async-std-1.10.0 | check | full | -1.15% | 5.77x
quote-stress | check | full | -1.09% | 5.44x
inotify-0.10.0 | check | full | -0.73% | 3.63x
time-macros-0.2.3 | check | full | -0.51% | 2.54x
*hyper-0.14.18 | check | full | -0.48% | 2.38x
*wg-grammar | check | full | -0.37% | 1.85x

File: compiler/rustc_expand/src/mbe/macro_rules.rs [hunk 1]
```diff
@@ -159,7 +159,7 @@ struct MacroRulesMacroExpander {
     name: Ident,
     span: Span,
     transparency: Transparency,
-    lhses: Vec<mbe::TokenTree>,
+    lhses: Vec<Result<Vec<MatcherLoc>, ()>>,
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:162]:
```suggestion
    lhses: Vec<Vec<MatcherLoc>>,
```
`Err`s here are treated as `span_bug`s by the logic below, so there should be no need to keep `Result`s here.

**Comment by petrochenkov** [comment]
>But the next commit will be moving `compute_locs`
outwards to a place that isn't suitable for the missing fragment
identifier checking

Why isn't it suitable?
`compile_declarative_macro` is actually a much better place for reporting these errors, because they are definition-site errors rather than use-site errors.
In `compile_declarative_macro` those errors can be reported right away, without stashing them to `ParseSess::missing_fragment_specifiers`, which then can be removed.

**Comment by petrochenkov** [comment]
I guess the next logical step is lowering macros to `MatcherLoc`s on encountering their definitions and also encoding them to metadata so other crates can load macros in already pre-compiled form, like other items.

File: compiler/rustc_expand/src/mbe/macro_rules.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_expand/src/mbe/macro_rules.rs:162]: Good point, I will fix this.

Comment (nnethercote) [reply compiler/rustc_expand/src/mbe/macro_rules.rs:162]: Oh, I remember why I did it this way. I was getting ICEs with malformed matchers like `x => (3)`. But I can handle that by only converting to `MatcherLoc` form in `compile_declarative_macro` if `valid` is true, and using an empty `Vec` instead.

**Comment by nnethercote** [comment]
> Why isn't it suitable? `compile_declarative_macro` is actually a much better place for reporting these errors, because they are definition-site errors rather than use-site errors. In `compile_declarative_macro` those errors can be reported right away, without stashing them to `ParseSess::missing_fragment_specifiers`, which then can be removed.

I wish this was true, because I would like to remove `missing_fragment_specifiers`. But the errors given depend on whether or not the macro with the missing fragment specifier is used. Consider this program:
```
#[allow(unused_macros)]
macro_rules! m_used { ( $id ) => {} }
macro_rules! m_unused { ( $id ) => {} }
fn main() { m_used!(x); }
```
Here's the current compiler output:
```
error: missing fragment specifier
 --> v.rs:2:25
  |
2 | macro_rules! m_used { ( $id ) => {} }
  |                         ^^^

error: missing fragment specifier
 --> v.rs:3:27
  |
3 | macro_rules! m_unused { ( $id ) => {} }
  |                           ^^^
  |
  = note: `#[deny(missing_fragment_specifier)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #40107 <https://github.com/rust-lang/rust/issues/40107>
```
Note that the two errors are different. Even worse, if you add `#![allow(missing_fragment_specifier)]`, then the second error disappears.

Based on this, if we want to keep the current behaviour, we need to keep the checking during macro expansion. If you think it's reasonable to change the behaviour here, I'd prefer to do it in a follow-up, to keep this PR just about performance improvements. Therefore, I have kept the checking in the macro expansion phase for now.

**Comment by petrochenkov** [comment]
@nnethercote 
The intent is to make
```rust
macro_rules! m_unused { ( $id ) => {} }
```
an error.

Previously it wasn't possible to catch missing specifiers at definition site without a significant rewrite, so the old implementation caught them at call site on a best effort basis.
It's ok to move it to the definition site in a separate PR.

**Comment by nnethercote** [comment]
> It's ok to move it to the definition site in a separate PR.

Thanks.

I just looked over the history here, these issues/PRs are relevant: #40107, #63239, #75516, #76605, #80296. Seems like this has been tried in the past and had to be reverted due to widespread breakage?

What steps are required for this? Would the `#[deny(missing_fragment_specifier)]` be removed? Would an MCP be needed? Would a crater run be needed? Can this only be done in a new edition?

**Comment by petrochenkov** [comment]
>Would the `#[deny(missing_fragment_specifier)]` be removed?

No, no, I suggest to keep it as a lint, as it is now, just report in at definition site (and even if the macro is unused).

**Pull request merged**